### PR TITLE
[util, gen_cheader] Generate additional multireg information

### PR DESCRIPTION
**Pull request by silvestrst**
Sometimes it is useful to know multireg fields width and "sub-register" information (fields per register and number of registers).

This is particularly useful in pinmux, and similar peripherals, where these values have to be calculated manually

NOTE: Some multireg registers could potentially have more then one field type, however at the moment only single filed cases are handled. Additional support if deemed useful, could always be added later.

It is worth to mention for the reviewers benefit that gen_cheader doesn't operate on "vanilla" IP `.hjson` file. Instead `validate.py` extends the parsed `.hjson ` by extrapolating some data found.

### CODE REVIEW

**Comment by silvestrst** [comment]
I think this is ready to land, @gkelly could you please have a look?

File: util/reggen/gen_cheader.py [hunk 1]
```diff
@@ -188,6 +205,43 @@ def gen_cdefines_module_params(outstr, module_data, module_name,
     genout(outstr, '\n')
 
 
+def gen_multireg_field_defines(outstr, regname, field, subreg_num, regwidth,
+                               existing_defines):
+    field_width = field['bitinfo'][1]
+    fields_per_reg = int(regwidth / field_width)
```
Comment (lenary) [inline util/reggen/gen_cheader.py:211]: This always rounds down, so you can just use the following for integer division:
```suggestion
    fields_per_reg = regwidth // field_width
```

File: util/reggen/gen_cheader.py [hunk 2]
```diff
@@ -188,6 +205,43 @@ def gen_cdefines_module_params(outstr, module_data, module_name,
...
+
+    define_name = regname + '_' + as_define(field['name'] + "_FIELD_WIDTH")
+    define = gen_define(define_name, [], str(field_width), existing_defines)
+    genout(outstr, define)
+
+    define_name = regname + '_' + as_define(field['name'] + "_FIELDS_PER_REG")
+    define = gen_define(define_name, [], str(fields_per_reg), existing_defines)
+    genout(outstr, define)
+
+    define_name = regname + '_' + as_define(field['name'] + "_REG_NUM")
```
Comment (lenary) [inline util/reggen/gen_cheader.py:221]: Do you think it might be clearer if the suffix here is `_MULTIREG_COUNT`?

Comment (silvestrst) [reply util/reggen/gen_cheader.py:221]: Sure, I will update it.

File: util/reggen/gen_cheader.py (hunk 1 above)
Comment (silvestrst) [reply util/reggen/gen_cheader.py:211]: Oh cool, thanks.

File: util/reggen/gen_cheader.py (hunk 2 above)
Comment (silvestrst) [reply util/reggen/gen_cheader.py:221]: I have also removed the field name from the count define, as it doesn't make any sense to have it there. So `RV_PLIC_IP_P_MULTIREG_COUNT` becomes `RV_PLIC_IP_MULTIREG_COUNT`.

File: util/reggen/gen_cheader.py (hunk 1 above)
Comment (silvestrst) [reply util/reggen/gen_cheader.py:211]: Done.

**Comment by silvestrst** [comment]
In the latest push:
* addressed all of the outstanding comments.

**Comment by lenary** [comment]
Ah, I've just seen #1876 which will affect this PR. Can you ensure you cope correctly with this kind of multireg?

**Comment by silvestrst** [comment]
@lenary good point I check it now in terms of my pinmux PR (it uses the new multireg defines).

**Comment by silvestrst** [comment]
As per offline discussion, the mentioned change should not have any impact as the new defines for now are only generated for a single field multireg. This is mentioned in the commit message. I will add the log entry for when generation of defines is skipped due to multiple fields in a multireg.

**Pull request merged**